src/c++/perf_analyzer/client_backend/triton_c_api/shared_memory_manager.cc

```diff
@@ -0,0 +1,184 @@
+}
+
+Error
+SharedMemoryManager::GetMemoryInfo(
+    const std::string& name, size_t offset, void** shm_mapped_addr,
+    TRITONSERVER_MemoryType* memory_type, int64_t* device_id)
+{
+  // protect shared_memory_map_ from concurrent access
+  std::lock_guard<std::mutex> lock(mu_);
+
+  auto it = shared_memory_map_.find(name);
+  if (it == shared_memory_map_.end()) {
+    return Error(
+        std::string("Unable to find shared memory region: '" + name + "'"));
+  }
+  if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+    *shm_mapped_addr =
+        (void*)((uint8_t*)it->second->mapped_addr_ + it->second->offset_ + offset);
+  } else {
+    *shm_mapped_addr = (void*)((uint8_t*)it->second->mapped_addr_ + offset);
+  }
+
+  *memory_type = it->second->kind_;
+  *device_id = it->second->device_id_;
+
+  return Error::Success;
+}
+
+
+Error
+SharedMemoryManager::Unregister(
+    const std::string& name, TRITONSERVER_MemoryType memory_type)
+{
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(mu_);
+
+  return UnregisterHelper(name, memory_type);
+}
+
+Error
+SharedMemoryManager::UnregisterAll(TRITONSERVER_MemoryType memory_type)
+{
+  std::lock_guard<std::mutex> lock(mu_);
+  std::string error_message = "Failed to unregister the following ";
+  std::vector<std::string> unregister_fails;
+  if (memory_type == TRITONSERVER_MEMORY_CPU) {
+    // Serialize all operations that write/read current shared memory regions
+    error_message += "system shared memory regions: ";
+    for (auto it = shared_memory_map_.cbegin(), next_it = it;
+         it != shared_memory_map_.cend(); it = next_it) {
+      ++next_it;
+      if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+        Error err = UnregisterHelper(it->first, memory_type);
+        if (!err.IsOk()) {
+          unregister_fails.push_back(it->first);
+        }
+      }
+    }
+  } else if (memory_type == TRITONSERVER_MEMORY_GPU) {
+    error_message += "cuda shared memory regions: ";
+    for (auto it = shared_memory_map_.cbegin(), next_it = it;
+         it != shared_memory_map_.cend(); it = next_it) {
+      ++next_it;
+      if (it->second->kind_ == TRITONSERVER_MEMORY_GPU) {
+        Error err = UnregisterHelper(it->first, memory_type);
+        if (!err.IsOk()) {
+          unregister_fails.push_back(it->first);
+        }
+      }
+    }
+  }
+
+  if (!unregister_fails.empty()) {
+    for (auto unreg_fail : unregister_fails) {
+      error_message += unreg_fail + " ,";
+    }
+    std::cerr << error_message << std::endl;
+    return Error(error_message);
+  }
+
+  return Error::Success;
+}
+
+Error
+SharedMemoryManager::UnregisterHelper(
+    const std::string& name, TRITONSERVER_MemoryType memory_type)
+{
+  // Must hold the lock on register_mu_ while calling this function.
+  auto it = shared_memory_map_.find(name);
+
+  // Remove region information from shared_memory_map_
+  shared_memory_map_.erase(it);
```

Comment:
I think `it` should be checked that it is not the `end()` iterator.

[https://en.cppreference.com/w/cpp/container/map/erase](https://en.cppreference.com/w/cpp/container/map/erase):

> Thus the end() iterator (which is valid, but is not dereferenceable) cannot be used as a value for pos.